### ISSUE TITLE
Option to stop training

### ISSUE THREAD

**Issue by fsmosca** [glinscott/nnue-pytorch#95]
Is there an option to automatically stop the training when the val loss of the last 5 or so epoch counter is no longer improving?

**Comment by kennyfrc** [comment]
@fsmosca You will have to set patience=5 using pytorch lightning's early stopping callback : https://pytorch-lightning.readthedocs.io/en/latest/common/early_stopping.html?highlight=early%20stoppage

**Comment by fsmosca** [comment]
Thanks, currently testing it.

**Comment by fsmosca** [comment]
This is [the branch](https://github.com/fsmosca/nnue-pytorch/commits/stop-training-bad-val-loss).

New options:

```
  --save-checkpoint-per-epoch
                        A flag to save a checkpoint with unique filename every time an epoch is completed.
  --save-checkpoint-epoch-end
                        A flag to save a checkpoint to a file last.ckpt every time an epoch is completed.
  --save-checkpoint-interval SAVE_CHECKPOINT_INTERVAL
                        An option to save checkpoint at epoch interval, default=1.
  --num-bad-val-loss-stop-training NUM_BAD_VAL_LOSS_STOP_TRAINING
                        Successive number of times the val loss of an epoch did not improve before the training is stopped, default=5.
```

### Early stopping
The last 2 epoch whose val_loss did not improve - training was stopped.

`--num-bad-val-loss-stop-training 2`

![](https://i.imgur.com/C1SR7x0.png)
